### ISSUE TITLE
Error: Cannot find module '../../third-party/specifications'

### ISSUE THREAD

**Issue by pintaric** [Brewskey/spark-server#98]
Is the dev branch currently broken? I just wanted to give the server a quick try after reading about it on community.particle.io, but I'm getting the following error message (see below). Am I doing something wrong? 

```
npm start

> spark-server@0.1.1 start /Users/Test/particle.io/spark-server
> nodemon --exec babel-node ./src/main.js --watch src --watch ../spark-protocol/dist --ignore data

[nodemon] 1.11.0
[nodemon] to restart at any time, enter `rs`
[nodemon] watching: /Users/Test/particle.io/spark-server/src/**/* ../spark-protocol/dist
[nodemon] starting `babel-node ./src/main.js`
module.js:472
    throw err;
    ^

Error: Cannot find module '../../third-party/specifications'
    at Function.Module._resolveFilename (module.js:470:15)
    at Function.Module._load (module.js:418:25)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/Test/particle.io/spark-server/node_modules/spark-protocol/dist/lib/FirmwareManager.js:67:23)
    at Module._compile (module.js:571:32)
    at Module._extensions..js (module.js:580:10)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/Test/particle.io/spark-server/node_modules/babel-register/lib/node.js:152:7)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/Test/particle.io/spark-server/node_modules/spark-protocol/dist/server/DeviceServer.js:63:24)
    at Module._compile (module.js:571:32)
    at Module._extensions..js (module.js:580:10)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/Test/particle.io/spark-server/node_modules/babel-register/lib/node.js:152:7)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/Test/particle.io/spark-server/node_modules/spark-protocol/dist/index.js:24:21)
    at Module._compile (module.js:571:32)
    at Module._extensions..js (module.js:580:10)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/Test/particle.io/spark-server/node_modules/babel-register/lib/node.js:152:7)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/Test/particle.io/spark-server/src/defaultBindings.js:5:1)
    at Module._compile (module.js:571:32)
    at loader (/Users/Test/particle.io/spark-server/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/Test/particle.io/spark-server/node_modules/babel-register/lib/node.js:154:7)
[nodemon] app crashed - waiting for file changes before starting...

```

**Comment by pintaric** [comment]
I should mention that I ran this on OSX using node v7.4.0 / npm v3.10.8.

**Comment by crenwick** [comment]
Same problem on node v6.9.4 / npm v4.1.1

**Comment by jlkalberer** [comment]
Looks like particle-cli just updated their code. We have a fix but are investigating long-term fixes.

**Comment by jlkalberer** [comment]
So, you'll need to delete `npm_modules` and `npm install`

That will run the postinstall script to fetch this file.

**Comment by crenwick** [comment]
You mean `node_modules` ? I've deleted that several times to rerun a fresh `npm install`, with no luck.

**Comment by jlkalberer** [comment]
Yeah, `node_modules`.

The only thing I can think of is that `spark-protocol` is being cached for
some reason. You can check the package and make sure this change is there -
https://github.com/Brewskey/spark-protocol/pull/47/files

**Comment by jlkalberer** [comment]
@crenwick @pintaric - is this issue solved for you two?

**Comment by crenwick** [comment]
I think so. Getting a few different errors on other tests, but several are now passing.

**Comment by jlkalberer** [comment]
Hmm.. the tests should all be passing. Can you let us know which ones are failing for you? A gist of the output would be good enough.

**Comment by pintaric** [comment]
It's now working for me, thanks for the quick response! :)

**Comment by crenwick** [comment]
Sure! https://gist.github.com/crenwick/282644be38fae360526757760a3a8922

This is with a fresh install of the npm modules.

**Comment by jlkalberer** [comment]
Ahh... ok, I have an idea of what's going on.

Did you change the path for the `FIRMWARE_REPOSITORY_DIRECTORY `?

Can you let me know what the path is for `spark-server` on your local machine?

**Comment by jlkalberer** [comment]
Sorry that you're the guinea pig.. We haven't tested on OSX or Linux and that's a brand new feature that's failing.

**Comment by pintaric** [comment]
I forgot to mention that I manually created the src/data directory and edited src/settings.js (to point FIRMWARE_REPOSITORY_DIRECTORY to the correct path on my machine) in order to get the server to run on OSX.

**Comment by jlkalberer** [comment]
@pintaric - are things working for you then? You just needed to manually create `src/data`?

**Comment by crenwick** [comment]
Oh interesting. Maybe I'm misunderstanding the scope of this project. I didn't know it was so coupled with firmware.

Is there a specific firmware repo that this requires? I pointed the `FIRMWARE_REPOSITORY_DIRECTORY` to my custom firmware repo, but its content is setup very different than what this project assumes (no `/user/applications` dir).

**Comment by pintaric** [comment]
As far as I know, the official firmware repo is located at [https://github.com/spark/firmware](https://github.com/spark/firmware).

**Comment by jlkalberer** [comment]
@crenwick - I added in the ability for the server to automatically flash system updates if your device needs an update.  This feature automatically downloads the necessary firmware updates from the `spark-firmware` releases during the post-install step.

The new issue you are seeing is coming from another feature I added.  You can use the particle CLI or the particle IDE to flash source directly to you. In order to do so, you need to clone the `https://github.com/spark/firmware` repository somewhere on your machine.  It will use this repo to compile you application.

Anyways, I'm going to write a readme on this that explains how to set the feature up.

**Comment by jlkalberer** [comment]
@pintaric - I'm talking about the local install path.  That's how you get cloud compile working.

**Comment by crenwick** [comment]
Sounds great, @jlkalberer 

Love the project.

**Comment by pintaric** [comment]
@jlkalberer - Thanks! Haven't gotten to that step yet, but I'll let you know if I run into any issues on OSX/Linux. Eventually, I'd like to run your spark-server fork in a Docker Container, so maybe that's something I can contribute in the future.

**Comment by jlkalberer** [comment]
@pintaric - pull latest and you shouldn't see any issues with running for the first time.

A docker container would be nice. I actually don't see there being any issues running it there.
